### ISSUE TITLE
Consider using binary format for data storage

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#577]
I wonder if gzipped xml format is significantly less efficient comparing to `file of` binary format in terms of loading time and CPU? Yes, gzipped binary's far more efficient in used disk space, but we don't care about that :D

Anyway, it would require recompilation of game data at every architecture change, less robust and (worst of all) requires additional work to be done.

See also https://sourceforge.net/p/castle-engine/discussion/general/thread/7cf74738/?limit=25#84d0

**Comment by michaliskambi** [comment]
Note that binary data can be platform-independent, if designed right :)

My long-term plan in CGE is to enable binary data for 3D/2D files using glTF format (sooner) and X3D binary encoding (maybe later) in Castle Game Engine.

Loading binary data can be indeed much faster in case of large arrays (e.g. large TVector3List for coordinates or TLongIntList for indexes). E.g. a huge array of TVector3List can then be loaded "directly" from disk to memory, without any CPU work on decompression/parsing/StrToFloat calls. Right now, loading an array of 1000 3D vertex positions means that we need to call StrToFloat 3000 times -- this is the main bottleneck, which would be avoided by loading binary data.

There are actually two approaches to "binary data" in glTF:

1. Use JSON to describe most mesh parameters, and the JSON file refers to external files for large binary data (like vertex coordinates), see e.g. https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Box/glTF-Draco . Each buffer can be either specified at text inside JSON, or be an external binary file.

2. Use 100% binary format, like https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Box/glTF-Binary .

My initial implementation would probably focus on 1., as it gives us 99% of the useful optimization, and I think (though I didn't check) it can be generated by Blender glTF exporter.

Once we have X3D binary encoding, we can even go one step further in the future, to extend https://github.com/castle-engine/castle-engine/wiki/Build-Tool to auto-generate binary versions of X3D files from their non-binary counterparts. This could work just like current "auto-generate-textures" works, https://github.com/castle-engine/castle-engine/wiki/Build-Tool#auto-generate-textures . This way you would have various options to use binary data:

- Change your asset workflow, to export glTF (already possible from Blender) or export binary X3D (unknown is this possible for any major 3D modeling software).
- Or don't change your asset workflow, instead use build tool's feature to call "auto-generate-binary-assets" before release. The engine would internally use the auto-generated binary version, if generated and up-to-date, otherwise engine could fallback on textual version.

**Comment by eugeneloza** [comment]
Yeah, that's exactly what this issue is about, but there is also a "huge-world" problem, which addresses not only X3D stuff loading, but also loading of maps, terrain, map objects, etc. E.g. just the `NavGrid` in the overworld itself is expected to be 30000x30000 (surely chunked and only partially loaded, but anyway, huge) and approximately the same amount of map objects. Storing that sort of things in XML might result in quiet poor performance.
